File: Prolog/exam_questions.pl
/************************************/
/****************2016****************/
/************************************/


/***************JANUARY**************/

noah([],[],[]).
noah([H1|T1],[H2|T2],[H1,H2|T3]):-
	noah(T1,T2,T3).

/***************AUGUST***************/

tr([[],[]],[]).

tr([[H1|T1],[H2|T2]],[[H1,H2]|Tail]):-
	tr([T1,T2],Tail).

tr([[H1,H2,H3]|Tail],[[H1|T1],[H2|T2],[H3|T3]]):-
	tr([T1,T2,T3],Tail).

File: Haskell/exam_questions.hs
import Data.List
import System.IO

---------------------------------------
-----------------2016------------------
---------------------------------------

----------------JANUARY----------------

revCount :: [a] -> [Int] -> [a]
revCount [] _ = []
revCount _ [] = []
revCount (x:xs) (y:ys) = revCount xs ys ++ replicate y x


----------------AUGUST----------------

tr :: [[a]] -> [[a]]
tr ([]:_) = []
tr a = (map head a) :  tr (map tail a)


---------------------------------------
-----------------2015------------------
---------------------------------------

----------------JANUARY----------------

afterFilter :: (a -> Bool) -> [a] -> [a]
afterFilter _ [] = []
afterFilter _ [_] = []
afterFilter f (x:y:xs)
  | f x = y : afterFilter f (y:xs)
  | otherwise = afterFilter f (y:xs)


----------------AUGUST----------------

revCount' :: [a] -> [Int] -> [a]
revCount' [] _ = []
revCount' _ [] = []
revCount' (x:xs) (y:ys) = (revCount' xs ys ++ replicate y x)


---------------------------------------
-----------------2014------------------
---------------------------------------

----------------JANUARY----------------

mapEveryOther :: (a -> a) -> [a] -> [a]
mapEveryOther _ [] = []
mapEveryOther f [x] = [f x]
mapEveryOther f (x:y:xs) = f x : y : mapEveryOther f xs


----------------AUGUST----------------

sorting' :: (a -> a -> Bool) -> [a] -> [a]
sorting' _ [] = []
sorting' p (n:ns) =
    let left = sorting' p (filter (p n) ns)
        right = sorting' p (filter (not . p n ) ns)
    in left ++ [n] ++ right 
    